### DISCUSSION TITLE
Difficulty consuming a C++ WinRT component in a Desktop App

### DISCUSSION THREAD

**Discussion by 4broadcast** [microsoft/CsWinRT#1881] [Q&A] [answered]
I have been following all of the canonical examples of projecting a simple C++ WinRT component into C#.  I am able to consume the component via project reference. And I can consume in a Console App via NuGet.

However, I have spent days trying to consume via NuGet in a "Desktop App." In this case, a  C# WinUI3 packaged app. I can't get it working in UWP either.

The issue I am running into is a Class Not Registered (0x80040154 (REGDB_E_CLASSNOTREG)) exception. I've been searching and searching GitHub & google to no avail. This exception is common, but none of the solutions I have found have been applicable, or just don't work. Any working example is based on project reference, not a NuGet package.

Does anyone have any guidance on getting past this roadblock? Is it even possible? I'm experimenting with developing some XAML controls in a C++ library, and would like to be able to distribute via NuGet package.

I can post a (working & non-working) sample if necessary.

Many Thanks - Jerry

**Comment by dongle-the-gadget** [comment]
For unpackaged apps, in your app manifest, you need to add `activatableClass` entries.

https://blogs.windows.com/windowsdeveloper/2019/04/30/enhancing-non-packaged-desktop-apps-using-windows-runtime-components/

**Comment by 4broadcast** [comment]
Hi And thank you, @dongle-the-gadget but I'm building packaged apps. I've read that blog a number of times, and even tried additing the entries, to no effect.

**Reply by dongle-the-gadget** [reply]
Oh. Is the WinMD present in the output folder?

**Reply by 4broadcast** [reply]
No, the WinMD was not present. None of the nuspecs I found included copying the WinMD for packages targeting "netx.0-windows10..."
![Screenshot 2024-12-04 124346](https://github.com/user-attachments/assets/557b4d3c-38d3-4894-a757-487dd020d7ba)

I'm having a lot of trouble finding online guidance on what files to include in the NuGet & where. 

Adding the WinMD to the NuGet pkg lib folder makes my class ambiguous: "The type 'SimpleMath' exists in both 'SampleComponent, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null, ContentType=WindowsRuntime' and 'SampleProjection, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'

I'm obviously at the limits of my understanding, so I really appreciate your help.

**Reply by dongle-the-gadget** [reply]
You should have the WinMD present next to your EXE output.

**Reply by 4broadcast** [reply]
I'm sorry, but I don't follow. Which winmd and what exe output? These are libraries, so I'm not creating an exe.

**Reply by manodasanW** [reply, marked as answer]
I have a feeling the issue here might be the `win10-x64` in the `runtimes` folder and that it should instead be `win-x64` due to the RID graph change in .NET.  But let me try this out locally and get back to you.  It is typically also recommended to package the winmd in the nuget package in another folder parallel to `lib` like a `winmd` folder, but typically the winmd is only needed when doing cross process calls which I believe isn't the case here.

**Reply by 4broadcast** [reply]
Thank you @manodasanW. That was exactly it - the RIDs are different, and  now it's `win-x64` . And it was staring me in the face for a week, if I had just cross-checked the .csproj, where the RID's are declared -> `<RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>`.

I doubt I'm done asking questions, but this gets me out of the weeds for a while.

**Comment by manodasanW** [comment]
What is the structure of your nuget package?  In addition to the winmd dongle asked about, is the C++/WinRT component DLL also in the output folder where the app exe is running from?

**Reply by 4broadcast** [reply]
Hi. No the dll is not getting copied. See my response to dongle for the package structure.

**Reply by gmgallo** [reply]
I struggled myself for weeks until I got a nuget package right. I used github copilot's help to create it. After a few iterations I made it work in C#. One other tip, the nupkg file is just a ZIP with another extension, if you have one working example rename it as .zip to inspect and duplicate its structure. In the nuspec and targs files make sure you have the directory levels right to navigate to the source and target folders.
